File: trijectory/engine/geometric_procedure.py
import numpy as np

from trijectory.type_aliases import ArrF64


def calc_relative_vector(r: ArrF64) -> ArrF64:  # (body, space) -> (body, body, space)
    # r = [[x0, y0],
    #      [x1, y1],
    #      [x2, y2]]
    return r[np.newaxis, :, :] - r[:, np.newaxis, :]


def calc_vectored_inv_square(r: ArrF64) -> ArrF64:  # (body, space) -> (body, body, space)
    rel = calc_relative_vector(r)
    non_zero_dist = np.linalg.norm(rel, axis=2) + np.eye(len(r))
    return rel / np.power(non_zero_dist[:, :, np.newaxis], 3)

File: src/rs_trijectory/geometric_procedure.rs
use ndarray::{Array2, Array3, Axis};
use ndarray_linalg::Norm;


pub fn calc_relative_vector(r: &Array2<f64>) -> Array3<f64> {
    // Equivalent of r[np.newaxis, :, :]
    let r_expanded_0 = r.view().insert_axis(Axis(0));
    // Equivalent of r[:, np.newaxis, :]
    let r_expanded_1 = r.view().insert_axis(Axis(1));

    // Broadcasting subtraction
    &r_expanded_0 - &r_expanded_1
}


pub fn calc_vectored_inv_square(r: &Array2<f64>) -> Array3<f64> {
    let rel = calc_relative_vector(r);
    let num_bodies = r.shape()[0];

    // Calculate the norm along the last axis (axis=2 in numpy)
    let non_zero_dist = rel.map_axis(Axis(2), |row| row.norm_l2());

    // Create an identity matrix and add it to non_zero_dist
    // This handles the np.eye(len(r)) part to avoid division by zero for self-interactions
    let identity_matrix = Array2::<f64>::eye(num_bodies);
    let non_zero_dist_with_identity = &non_zero_dist + &identity_matrix;

    // Equivalent of np.power(non_zero_dist[:, :, np.newaxis], 3)
    let powered_dist = non_zero_dist_with_identity
        .insert_axis(Axis(2))
        .mapv(|x| x.powi(3));

    // Equivalent of rel / powered_dist
    rel / powered_dist
}
